### PR TITLE
new resource `azurerm_stack_hci_extension`

### PR DESCRIPTION
**Pull request by teowa**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

new resource `azurerm_stack_hci_extension`
azure doc: https://learn.microsoft.com/en-us/azure-stack/hci/manage/arc-extension-management?tabs=azureportal
rest api: https://github.com/Azure/azure-rest-api-specs/blob/c7b98b36e4023331545051284d8500adf98f02fe/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/stable/2024-01-01/extensions.json

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [ ] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

Tests depends on Stack HCI Arc Setting resource generated during HCI deployment, the HCI deployment steps is included in PR https://github.com/hashicorp/terraform-provider-azurerm/pull/25646, my local tests can pass with a existing Arc Setting resource.

```
ARM_TEST_STACK_HCI_ARC_SETTING_ID='/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/acctest-hci-0804/providers/Microsoft.AzureStackHCI/clusters/hci0804-cl/arcSettings/default' make acctests SERVICE="azurestackhci" TESTARGS="-parallel 5 -run TestAccStackHCIExtension_" TESTTIMEOUT=2h
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/azurestackhci -parallel 5 -run TestAccStackHCIExtension_ -timeout 2h -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccStackHCIExtension_basic
=== PAUSE TestAccStackHCIExtension_basic
=== RUN   TestAccStackHCIExtension_complete
=== PAUSE TestAccStackHCIExtension_complete
=== RUN   TestAccStackHCIExtension_update
=== PAUSE TestAccStackHCIExtension_update
=== RUN   TestAccStackHCIExtension_requiresImport
=== PAUSE TestAccStackHCIExtension_requiresImport
=== CONT  TestAccStackHCIExtension_basic
=== CONT  TestAccStackHCIExtension_requiresImport
=== CONT  TestAccStackHCIExtension_update
=== CONT  TestAccStackHCIExtension_complete
--- PASS: TestAccStackHCIExtension_requiresImport (173.50s)
--- PASS: TestAccStackHCIExtension_complete (175.84s)
--- PASS: TestAccStackHCIExtension_basic (176.22s)
--- PASS: TestAccStackHCIExtension_update (400.56s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/azurestackhci 400.578s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* new resource `azurerm_stack_hci_extension`

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/azurestackhci/stack_hci_extension_resource_test.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_stack_hci_extension" "test" {
+  name           = "acctest-shce-%[2]s"
+  arc_setting_id = %[3]q
+  publisher      = "Microsoft.EnterpriseCloud.Monitoring"
+  type           = "MicrosoftMonitoringAgent"
+
+  protected_settings = <<PROTECTED_SETTINGS
```
Comment (ms-zhenhua) [inline internal/services/azurestackhci/stack_hci_extension_resource_test.go:140]: `basic` should not contain optional properties: `protected_settings` and `settings`

File: website/docs/r/stack_hci_extension.html.markdown [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+* `name` - (Required) The name which should be used for this Azure Stack HCI Extension. Changing this forces a new resource to be created.
+
+* `arc_setting_id` - (Required) The ID of the Azure Stack HCI Cluster Arc Setting. Changing this forces a new resource to be created.
+
+* `publisher` - (Required) The name of the extension handler publisher, such as `Microsoft.Azure.Monitor`. Changing this forces a new resource to be created.
+
+* `type` - (Required) Specifies the type of the extension. For example `CustomScriptExtension` or `AzureMonitorLinuxAgent`. Changing this forces a new resource to be created.
+
+* `auto_upgrade_minor_version_enabled` - (Optional) Indicates whether the extension should use a newer minor version if one is available at deployment time. Once deployed, however, the extension will not upgrade minor versions unless redeployed, even with this property set to true. Possible values are `true` and `false`.
```
Comment (ms-zhenhua) [inline website/docs/r/stack_hci_extension.html.markdown:64]: what are the differences between `auto_upgrade_minor_version_enabled` and `automatic_upgrade_enabled`?

File: website/docs/r/stack_hci_extension.html.markdown [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+* `settings` - (Optional) The json formatted public settings for the extension.
+
+* `type_handler_version` - (Optional) Specifies the version of the script handler.
+
+-> **NOTE:** `type_handler_version` cannot be set when `automatic_upgrade_enabled` is set to `true`.
+
+-> **NOTE:** Possible values for `type_handler_version` can be found using the Azure CLI, e.g.:
+
+```shell
+az vm extension image list --publisher Microsoft.Azure.Monitor -n AzureMonitorWindowsAgent --location westus -o table
```
Comment (ms-zhenhua) [inline website/docs/r/stack_hci_extension.html.markdown:79]: Shall we provide a link to this Azure CLI instead of directly providing the shell command?

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @teowa, 

Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	TypeHandlerVersion             string `tfschema:"type_handler_version"`
+}
+
+func (r StackHCIExtensionResource) Arguments() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: validation.All(
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:54]: no length limitation found from API testing or doc

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		},
+
+		"type": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"auto_upgrade_minor_version_enabled": {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:81]: added default:true

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+		},
+
+		"automatic_upgrade_enabled": {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:86]: added default:true

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+			if err != nil {
+				return err
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("%s was not found", id)
+				}
+				return fmt.Errorf("reading %s: %+v", id, err)
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:221]: changed

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+			}
+
+			// protected_settingss is not returned in the response, so we read it from the state
+			var extension StackHCIExtensionResourceModel
+			if err := metadata.Decode(&extension); err != nil {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:226]: fixed

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			model := resp.Model
+			if model == nil || model.Properties == nil || model.Properties.ExtensionParameters == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", id)
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:303]: fixed

File: internal/services/azurestackhci/stack_hci_extension_resource.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+				}
+
+				model.Properties.ExtensionParameters.Settings = pointer.To(interface{}(expandedSetting))
+			}
+
+			if metadata.ResourceData.HasChange("type_handler_version") {
+				model.Properties.ExtensionParameters.TypeHandlerVersion = pointer.To(config.TypeHandlerVersion)
+			}
+
+			if err := client.CreateThenPoll(ctx, *id, *model); err != nil {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:336]: fixed

File: internal/services/azurestackhci/stack_hci_extension_resource_test.go (hunk 1 above)
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource_test.go:140]: fixed

File: website/docs/r/stack_hci_extension.html.markdown (hunk 3 above)
Comment (teowa) [reply website/docs/r/stack_hci_extension.html.markdown]: Removed because this cannot be found in doc, and it might be inaccurate.

**Comment by teowa** [comment]
Hi @ms-zhenhua , thanks for reviewing this, please kindly take another look.

```
ARM_TEST_LOCATION=australiaeast ARM_TEST_STACK_HCI_ARC_SETTING_ID='/subscriptions/XXX/resourceGroups/acctest-hci-0812/providers/Microsoft.AzureStackHCI/clusters/hci0812-cl/arcSettings/default' make acctests SERVICE="azurestackhci" TESTARGS="-parallel 5 -run TestAccStackHCIExtension" TESTTIMEOUT=2h

==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/azurestackhci -parallel 5 -run TestAccStackHCIExtension -timeout 2h -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccStackHCIExtension_basic
=== PAUSE TestAccStackHCIExtension_basic
=== RUN   TestAccStackHCIExtension_complete
=== PAUSE TestAccStackHCIExtension_complete
=== RUN   TestAccStackHCIExtension_update
=== PAUSE TestAccStackHCIExtension_update
=== RUN   TestAccStackHCIExtension_requiresImport
=== PAUSE TestAccStackHCIExtension_requiresImport
=== CONT  TestAccStackHCIExtension_basic
=== CONT  TestAccStackHCIExtension_update
=== CONT  TestAccStackHCIExtension_requiresImport
=== CONT  TestAccStackHCIExtension_complete
--- PASS: TestAccStackHCIExtension_basic (120.85s)
--- PASS: TestAccStackHCIExtension_requiresImport (196.44s)
--- PASS: TestAccStackHCIExtension_complete (239.01s)
--- PASS: TestAccStackHCIExtension_update (251.42s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/azurestackhci 251.451s
```

File: internal/services/azurestackhci/stack_hci_extension_resource.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ms-zhenhua) [reply internal/services/azurestackhci/stack_hci_extension_resource.go:224]: seems not changed?

[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/azurestackhci/stack_hci_extension_resource.go]: sorry for missing this one, updated

**Review by ms-zhenhua — APPROVED** [review]
Hi @teowa ,

Thank you for your updates. LGTM~

**Comment by katbyte** [comment]
@teowa - could we fix up the merge conflicts please?

**Review by katbyte — APPROVED** [review]
LGTM 🌳

**Pull request merged**